sycl/doc/extensions/proposed/sycl_ext_oneapi_range_type.asciidoc

```diff
@@ -0,0 +1,168 @@
+
+* link:../experimental/sycl_ext_oneapi_kernel_properties.asciidoc[
+  sycl_ext_oneapi_kernel_properties]
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+The maximum number of work-items that can be launched in a single kernel
+depends on multiple factors.
+SYCL 2020 says that the total number of work-items must be representable as a
+`size_t`, but several implementations (including {dpcpp}) provide optimization
+options to assert that kernels will not require the full range of a `size_t`.
+
+This extension proposes a new kernel property that allows developers to declare
+the range requirements of individual kernels, providing more fine-grained
+control than existing compiler options and improved error behavior.
+
+The property described in this extension is an advanced feature that most
+applications should not need to use.
+In most cases, applications get the best performance without using this
+property.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_RANGE_TYPE` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New kernel property
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+struct range_type_key {
+  template <typename T>
+  using value_t = property_value<range_type_key, T>;
+};
+
+template <typename T>
+inline constexpr range_type_key::value_t<T> range_type;
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+|===
+|Property|Description
+
+|`range_type`
+|The `range_type` property adds the requirement that the kernel must be
+compatible with kernel launches where the linear index of a work-item lies
+in the range [0, `std::numeric_limits<T>::max()`).
+
+If the implementation cannot satisfy this requirement, the implementation
+must throw an `exception` with the `errc::kernel_not_supported` error code,
+regardless of launch configuration.
+If the implementation can satisfy this requirement, but the kernel is
+launched with an incompatible configuration, the implementation must throw
+an `exception` with the `errc::nd_range` error code.
```

Comment:
Can we word it like this instead?

> The `range_type` property is an assertion by the application that the kernel will never be launched with more than `std::numeric_limits<T>::max()` work-items.  If the kernel is launched with more than this many work-items, the implementation must throw a synchronous `exception` with the `errc::nd_range` error code.

It seems like you are going out of your way to add more wiggle room for the implementation to fail, so I suspect you have some cases in mind that will be problematic?

Reply:
I was trying to distinguish between cases where the implementation doesn't support the requested range type and cases where the user requested too many work-items. For example, a kernel with `range_type<size_t>` would throw `errc::kernel_not_supported` on any device with a hard limit of a 32-bit range for work-items, even if supplied `nd_range` would have fit in the 32-bit range.

To see why this is useful, here are some use-cases I had in mind when I sketched this out:
- `-fsycl-id-queries-fit-in-int` + `range_type<size_t>` => Compile specified kernels with 64-bit support (or fail)
- `-fno-sycl-id-queries-fit-in-int` + `range_type<int>` => Compile specified kernels as-if id-queries-fit-in-int
- `range_type<unsigned short>` => Kernel assumes a maximum of 65535 work-items

The third use-case is covered by your wording, but I don't think the first two are.

Reply:
It seems like it might be better for `range_type<T>` to override `-fsycl-id-queries-fit-in-int` rather than vice versa.  We could specify `-fsycl-id-queries-fit-in-int` as adding an implied `range_type<int>` to every kernel that does not already have the `range_type` property.  I think we have this pattern elsewhere in DPC++, for example with the option that sets the default sub-group size.

The specification of `range_type` could then be as I describe above.  The compiler would use this information to reduce the number of bits it needs to store the linear id.

Reply:
That's a good point.  I was thinking that it would be too hard to change the behavior of the existing flag, but it would be much cleaner that way.  And we'd still get the behavior that I really want, which is that when `range_type<size_t>` is present the compiler _must_ support a 64-bit range, and when `range_type<int>` is present, the compiler _may_ optimize for a 32-bit range.

You've convinced me, so I've adopted your proposed wording in c53b75bc747ea02d4d6f04290d400c0495ebb7ab.

Reply:
I think this looks good, and I've appoved.

You might consider capturing the interaction with the `fsycl-id-queries-fit-in-int` options in a non-normative section.  When we do this, we add a section that is a peer of "Specification" titled something like:

```
== Interaction with the {dpcpp} "-fsycl-id-queries-fit-in-int" option

The `-fsycl-id-queries-fit-in-int` option is specific to the
{dpcpp} implementation.
Therefore, this section that describes the interaction between this extension
and that option is non-normative and does not apply to other SYCL
implementations that may support this extension.
```

Reply:
Good idea.  I've done that in 4d248cb2b6c25d7e61537c8cc6993afc7f8ca3ad, borrowing some of the language from the default sub-group size extension.